tests/data_context/store/test_expectations_store.py

```diff
@@ -310,8 +310,177 @@ def test_get_key_in_non_cloud_mode(empty_data_context):
+    expectation = ExpectColumnValuesToBeInSet(
+        id=provided_id,
+        column="a",
+        value_set=[1, 2, 3],
+        result_format="BASIC",
+    )
+
+    # Act
+    store.add_expectation(suite=suite, expectation=expectation)
+
+    # Assert
+    updated_suite_dict = store.get(key=key)
+    updated_suite = ExpectationSuite(**updated_suite_dict)
+    added_expectation = updated_suite.expectations[0]
+    assert UUID(added_expectation.id)
+    assert added_expectation.id != provided_id
+
+
+@pytest.mark.cloud
+def test_update_expectation_success_cloud_backend(empty_cloud_data_context):
+    # Arrange
+    context = empty_cloud_data_context
+    store = context.expectations_store
+    suite_name = "test-suite"
+    expectation = ExpectColumnValuesToBeInSet(
+        column="a",
+        value_set=[1, 2, 3],
+        result_format="BASIC",
+    )
+    suite = context.add_expectation_suite(
+        suite_name, expectations=[expectation.configuration]
+    )
+    key = store.get_key(suite)
+    expectation = suite.expectations[0]
+    updated_column_name = "foo"
+    expectation.column = updated_column_name
+
+    # Act
+    store.update_expectation(suite=suite, expectation=expectation)
+
+    # Assert
+    updated_suite_dict = store.get(key=key)
+    updated_suite = ExpectationSuite(**updated_suite_dict)
+    updated_expectation = updated_suite.expectations[0]
+    assert updated_expectation.id == expectation.id
+    assert updated_expectation.column == updated_column_name
+
+
+@pytest.mark.filesystem
+def test_update_expectation_success_file_backend(empty_data_context):
+    # Arrange
+    context = empty_data_context
+    store = context.expectations_store
+    suite_name = "test-suite"
+    expectation = ExpectColumnValuesToBeInSet(
+        column="a",
+        value_set=[1, 2, 3],
+        result_format="BASIC",
+    )
+    suite = context.add_expectation_suite(
+        suite_name, expectations=[expectation.configuration]
+    )
+    key = store.get_key(suite)
+    expectation = suite.expectations[0]
+    updated_column_name = "foo"
+    expectation.column = updated_column_name
+
+    # Act
+    store.update_expectation(suite=suite, expectation=expectation)
+
+    # Assert
+    updated_suite_dict = store.get(key=key)
+    updated_suite = ExpectationSuite(**updated_suite_dict)
+    updated_expectation = updated_suite.expectations[0]
+    assert updated_expectation.id == expectation.id
+    assert updated_expectation.column == updated_column_name
+
+
+@pytest.mark.filesystem
+def test_update_expectation_raises_error_for_missing_expectation(empty_data_context):
+    # Arrange
+    context = empty_data_context
+    store = context.expectations_store
+    suite_name = "test-suite"
+    expectation = ExpectColumnValuesToBeInSet(
+        id="2b284004-0e0e-455d-a7f4-11e162fd06c9",
+        column="a",
+        value_set=[1, 2, 3],
+        result_format="BASIC",
+    )
+    suite = context.add_expectation_suite(
```

Comment:
This seems to be a strange test because it's basically demonstrating `add_expectation_suite` behaves in an non-intuitive way. That is, adding the `expectation.configuration` doesn't add the expectation. We don't want to preserve this behavior and we want to get rid of `add_expectation_suite` where we pass in configuration instead of Expectation.

While this documents current behavior, I would delete this test since this isn't guarding behavior we want and we plant to remove this before we ship. Another option would be to add a comment explaining what this is testing and why.

If I'm refactoring and see I test, I usually assume that it's behavior I want to preserve.

Reply:
hm good catch - this test wasn't supposed to pass an expectation in here, so i'm surprised that it passes.

Reply:
ah right - the issue here is that the ID which was passed in with the expectation gets overwritten, since the backend defines IDs. So when we try to update that expectation, no IDs match and it raises an error. I'll remove the expectation from the suite's init to make that clearer.

I do see why this behavior was unexpected as a reader of the code, though - should we raise an exception instead of overwriting IDs?

Reply:
updated in [b261708](https://github.com/great-expectations/great_expectations/pull/9125/commits/b26170837df1d346a0633c33950306eaa7ef47b9)